experimental/grpc_attestation/src/attestation.rs

```diff
@@ -0,0 +1,210 @@
+}
+
+impl Receiver {
+    /// Receives requests from [`Receiver::request_stream`].
+    /// `Ok(None)` output means that the corresponding gRPC stream has ended.
+    async fn receive(&mut self) -> anyhow::Result<Option<GetAttestedDataRequest>> {
+        let request = if let Some(request) = self.request_stream.next().await {
+            Some(request.context("Couldn't receive request")?)
+        } else {
+            None
+        };
+        Ok(request)
+    }
+}
+
+struct RequestHandler {
+    // Decrypts and encrypts messages using a Diffie-Hellman session key.
+    encryptor: AeadEncryptor,
+    // Processes data from client requests and creates responses.
+    handler: fn(&[u8]) -> Vec<u8>,
+}
+
+impl RequestHandler {
+    pub fn new(encryptor: AeadEncryptor, handler: fn(&[u8]) -> Vec<u8>) -> Self {
+        Self { encryptor, handler }
+    }
+
+    /// Decrypts a client request, runs [`RequestHandler::handler`] on them and returns an encrypted
+    /// response.
+    pub async fn handle_request(
+        &mut self,
+        receiver: &mut Receiver,
+    ) -> anyhow::Result<Option<GetAttestedDataResponse>> {
+        if let Some(request) = receiver
+            .receive()
+            .await
+            .context("Couldn't receive request")?
+        {
+            let request_type = request.request_type.context("Couldn't read request type")?;
+            let data = if let RequestType::Data(data) = request_type {
+                Ok(data)
+            } else {
+                Err(anyhow!("Received incorrect message type"))
+            }?;
+
+            let decrypted_data = self
+                .encryptor
+                .decrypt(&data)
+                .context("Couldn't decrypt data")?;
+            let response = (self.handler)(&decrypted_data);
+
+            let encrypted_data = self
+                .encryptor
+                .encrypt(&response)
+                .context("Couldn't encrypt data")?;
+            let encrypted_response = GetAttestedDataResponse {
+                response_type: Some(ResponseType::Data(encrypted_data)),
+            };
+
+            Ok(Some(encrypted_response))
+        } else {
+            Ok(None)
+        }
+    }
+}
+
+/// gRPC Attestation Service implementation.
+pub struct AttestationServer {
+    /// PEM encoded X.509 certificate that signs TEE firmware key.
+    tee_certificate: Vec<u8>,
+    /// Processes data from client requests and creates responses.
+    request_handler: fn(&[u8]) -> Vec<u8>,
+}
+
+impl AttestationServer {
+    pub fn create(
+        tee_certificate: Vec<u8>,
+        request_handler: fn(&[u8]) -> Vec<u8>,
+    ) -> anyhow::Result<Self> {
+        Ok(Self {
+            tee_certificate,
+            request_handler,
+        })
+    }
+
+    /// Attest a single gRPC streaming request. Client messages are provided via `receiver`.
+    async fn attest(
+        receiver: &mut Receiver,
+        tee_certificate: Vec<u8>,
```

Comment:
Why is this method not defined on &self directly, and use the cert from there?

Reply:
Because it is used inside a stream implementation, which this class implements:
```rust
        let response_stream = async_stream::try_stream! {
            let mut receiver = Receiver { request_stream };
            let (attestation_response, encryptor) = Self::attest(&mut receiver, tee_certificate)
                .await
                .map_err(Status::internal("Couldn't attest to the client"))?;
```